### PR TITLE
Improve placement of `use` suggestions

### LINKED ISSUE #42548: Weird help message suggest to place `use` randomly, when a macro is called in scope
Repro case (1.19.0-beta.1 as well as nightly):

```rust
macro_rules! y {
    () => {}
}
mod m {
    pub const A: i32 = 0;
}
fn main() {
    y!();
    let _ = A;
}
```

When running it in nightly, it produces a weird error message:

```
error[E0425]: cannot find value `A` in this scope
 --> 1.rs:9:13
  |
9 |     let _ = A;
  |             ^ not found in this scope
  |
help: possible candidate is found in another module, you can import it into scope
  | fn main() use m::A;

error: aborting due to previous error(s)
```

<details><summary>No repro on 1.18 or before</summary>

```
error[E0425]: cannot find value `A` in this scope
 --> <source>:9:13
  |
9 |     let _ = A;
  |             ^ not found in this scope
  |
  = help: possible candidate is found in another module, you can import it into scope:
            `use m::A;`
error: aborting due to previous error
```

</details>

<details><summary>Version</summary>

```
rustc 1.19.0-nightly (f062832b2 2017-06-07)
binary: rustc
commit-hash: f062832b208e94f2f0f26ed7fb5c48c172069fbe
commit-date: 2017-06-07
host: x86_64-apple-darwin
release: 1.19.0-nightly
LLVM version: 4.0
```

</details>

### LINKED ISSUE #42835: Import suggestions suggest inserting imports in the wrong place
For this kind of error:

```
pub mod foo {
    struct Foo;
}

fn main() {
    let a = Foo;
}
```

We get an error like

```
error[E0425]: cannot find value `Foo` in this scope
  --> <anon>:10:14
   |
10 |     let a = Foo;
   |             ^^^ not found in this scope
   |
help: possible candidate is found in another module, you can import it into scope
   | use foo::Foo;
```

Which is good. But if you put this whole example in a submodule and then look at the json error, the suggestion is add the import to the enclosing module, not the submodule, which is incorrect.

cc #42823, @oli-obk

### LINKED ISSUE #43769: `use` suggestions corrupted w/ nearby code when inside items
This is... bizarre.

### STR

```rust
fn main() {
    type Dict<K, V> = HashMap<K, V>;
}
```

### Expected

```
help: possible candidates are found in other modules, you can import them into scope
  | use std::collections::HashMap;
  | use std::collections::hash_map::HashMap;
```

### Actual

```
help: possible candidates are found in other modules, you can import them into scope
  | fn main() use std::collections::HashMap;
  | fn main() use std::collections::hash_map::HashMap;
```

<br/>
The issue seems to arise from the erroneous code being inside another block. I originally encountered this with code generated by `lazy_static!`.

### PR DESCRIPTION
**Pull request by oli-obk**
r? @nrc

cc @estebank @Mark-Simulacrum 

fixes #42835
fixes #42548
fixes #43769

### CODE REVIEW

File: src/librustc_resolve/lib.rs [hunk 1]
```diff
@@ -2530,10 +2589,14 @@ impl<'a> Resolver<'a> {
...
-            err
+            (err, candidates)
         };
         let report_errors = |this: &mut Self, def: Option<Def>| {
-            report_errors(this, def).emit();
+            let (err, candidates) = report_errors(this, def);
+            let def_id = this.current_module.normal_ancestor_id;
+            let node_id = this.definitions.as_local_node_id(def_id).unwrap();
+            let better = def.is_some();
+            this.use_injections.push(UseError{ err, candidates, node_id, better });
```
Comment (estebank) [inline src/librustc_resolve/lib.rs]: Space between `UseError` and `{`?

File: src/librustc_resolve/lib.rs [hunk 2]
```diff
@@ -581,6 +581,53 @@ impl<T> ::std::ops::IndexMut<Namespace> for PerNS<T> {
...
+}
+
+impl<'tcx> Visitor<'tcx> for UsePlacementFinder {
+    fn visit_mod(
+        &mut self,
+        module: &'tcx ast::Mod,
+        _: Span,
+        _: &[ast::Attribute],
+        node_id: NodeId,
+    ) {
```
Comment (estebank) [inline src/librustc_resolve/lib.rs]: Isn't the following slightly easier to follow?

```rust
    fn visit_mod(&mut self,
                 module: &'tcx ast::Mod,
                 _: Span,
                 _: &[ast::Attribute],
                 node_id: NodeId) {
```

**Review by estebank — APPROVED** [review]
Love it! The new locations are exactly where they should be. The only wart I see is that by applying these suggestions without any smarts the resulting output would be the following:

```rust
use a::A;
struct S;
```

I don't think we should change the compiler output, I think clients could perform local reformatting to comply with `rustfmt` instead.

File: src/librustc_resolve/lib.rs (hunk 2 above)
Comment (oli-obk) [inline src/librustc_resolve/lib.rs]: Rustfmt has my version as the new style and I prefer it, too, but I can change it to the other style to fit more into what's still more prominent in rustc.

**Comment by oli-obk** [comment]
> The only wart I see is that by applying these suggestions without any smarts the resulting output would be the following:
> ```
> use a::A;
> struct S;
> ```
> I don't think we should change the compiler output, I think clients could perform local reformatting to comply with rustfmt instead.

I'm confused... so you want the final output to be `use a::A;struct S;`?

**Comment by estebank** [comment]
I'd want the final output to be

’’’rust
use a::A;

struct S;
’’’

But wouldn't change rustc for it, as it makes the cli output worse.

**Comment by oli-obk** [comment]
> But wouldn't change rustc for it, as it makes the cli output worse.

Oh I already have a hack that makes sure the cli is fine ;) Adding another newline to the suggestion output in case the following item is not a `use` would definitely be possible.

**Pull request merged**